Reprenez le contrôle de vos file system avec Flysystem (#708)
eleven-labs/blog.eleven-labs.com: _posts/fr/2019-11-20-reprenez-le-controle-de-vos-file-system-avec-flysystem.md

---
layout: post

> title: Reprenez le contrôle de vos file system avec Flysystem

Reviewer: On voit le voit plus souvent la forme "filesystem" mais "file system" reste valide.

excerpt: Cet article vous présente comment interagir simplement avec vos systèmes de gestion de fichier avec Flysystem.
authors:
    - nicolas
lang: fr
permalink: /fr/reprenez-le-controle-de-vos-file-system-avec-Flysystem/
categories:
    - PHP
    - Bundle
    - File system
tags:
    - PHP
    - Symfony
    - Flysystem
    - File system
---

> # Introduction

Reviewer: `## Introduction` Utilisé ici un titre de niveau 2 car le niveau 1 est utilisé par le titre de ton article une fois sur le blog :)

> La gestion d’un ou plusieurs système de fichier dans une application peut s’avérer compliqué. Si on utilise une solution distante comme un stockage en cloud ou sur un serveur SFTP il faudra s’occuper de la connection. Et dans tous les cas il faudra créer un système  d’interaction avec nos ressources pour les différentes solution que l’on pourra réutiliser dans toute notre application.

Reviewer: Personnellement, je commencerai par dire que manipuler un système de fichier depuis PHP n'est pas trivial, je montrerai comment le faire sur un système de fichier local puis j'élargirai le propos de la façon suivante: En substance, "nous avons utilisé les fonctions natives de PHP pour manipuler un système de fichier local, mais que faire si l'ont me demande de changer de solution pour la gestion de mes fichiers ? Par exemple, les stocker et les manipuler sur Amazon S3 ? Je tire la lib AWS PHP, je la configure, change mon code radicalement et boom ça marche. Sauf que demain, nous passons sur Azure et tous nos fichiers sont maintenant stockés sur Azure Blob Storage. Pétage de câble." 

Parler du patron Adaptateur (https://fr.wikipedia.org/wiki/Adaptateur_(patron_de_conception)) qui permet d'abstraire les grandes opérations de manipulation de fichier via une interface puis présenter FlySystem

Pour gérer la connection avec les solutions de stockage en cloud nous pouvons trouver un sdk qui nous simplifiera notre notre code. Seulement pour l'interaction avec nos resource nous seront obligé de créer ce genre de classe :

> ![Class CRUD AWS S3]({{ site.baseurl }}/assets/2019-11-20-reprenez-le-controle-de-vos-file-system-avec-flysystem/screenshot-class.png){: style="margin: 0 auto; display: block;" }

Reviewer: L'image est difficilement visible sur le site, si l'intention est de démontrer que maintenir soi même une abstraction du système de fichier et les différentes implémentations qui en découle (local, memory, AWS S3 et autres) sans parler des test d'intégrations qu'il faudra écrire et maintenir et une tâche complexe, le plus simple et encore de l'écrire.

![Method delete AWS S3]({{ site.baseurl }}/assets/2019-11-20-reprenez-le-controle-de-vos-file-system-avec-flysystem/screenshot-method.png){: style="margin: 0 auto; display: block;" }

Ca fait du code à maintenir et si un jour on l’on doit changer de solution de stockage l’on peut avoir des problématiques de compatibilité.

 # Voici la solution que je vous propose

La solution est donc d'utiliser une librairie qui vas faire le travail pour vous, Flysystem. 
[Flysystem](https://flysystem.thephpleague.com/docs/) développer par [thephpleague](https://thephpleague.com/fr/), un groupe de développer de bibliothèques PHP. *Flysystem* est une bibliothèque d'abstraction du système de fichier. 
Cela permet donc de changer du solution de système de fichier rapidement et facilement. 
Vous pouvez l’utiliser dans une application PHP avec ou sans framework.

Flysystem fourni une API permettant de gérer vos ressources sur un grand nombre de système de fichier. D’office, la librairie fournit trois adaptateurs de système de fichier, FTP, Local et NullAdapter. Mais rien ne vous empêche d’ajouter d’autre adaptateur de système de fichier, d'ailleurs il en existe un grand nombre.
 

> Voici la liste des adaptateurs développés par The PHP league :

Reviewer: Je ne pense pas qu'il soit nécessaire dans faire la liste tu peux en cité quelques uns dans un paragraphe et filer le lien vers la docs.

AWS S3 V2 et V3
Azure Blob Storage
Menory
PHPCR
Rackspace Cloud Files
SFTP
WebDAV
Zip

> Et quelque uns des adaptateurs développer par la communauté :

Reviewer: Même commentaire que ci dessus

      adapter: 'aws'
      options:
        client: Aws\S3\S3Client
        bucket: '%env(string:AWS_BUCKET)%'
```

Alors dans c’est exemple je nomme mes systèmes de fichier par rapport à leurs adaptateur mais vous pouvez les nommer par rapport à leur utilité exemple :

```yaml
flysystem:
    assets.storage:
      adapter: 'local'
      # ...

    invoice.storage:
      adapter: 'gcloud'
      # ...

    media.storage:
      adapter: 'aws'
      # ...
``` 

Et voilà pour la configuration de plusieurs système de fichier.

# La configuration ça va deux minutes

Nous avons vu comment permuter de système de fichier juste avec quelque lignes de configuration, mais aussi comment configurer plusieurs système de fichier.
Maintenant nous allons voir comment utiliser nos système de fichier. [l’API Flysystem](https://flysystem.thephpleague.com/docs/usage/filesystem-api/) nous offre plusieurs méthode pour manipuler des resources. Voici la liste des méthodes 

- write
- writeStream
- update
- updateStream
- put
- putStream
- read
- readStream
- has
- delete
- readAndDelete
- rename
- copy
- getMimetype
- getTimestamp
- getSize
- createDir
- deleteDir
- listContents
- setVisibility

> - addPlugin

Reviewer: Alors tu peux donner la signature de l'interface mais garde à l'esprit que certaines méthodes ne sont pas triviales à comprendre genre `getVisibility`, `addPlugin` et toutes les méthodes `*Stream`. Coller l'interface PHP dans le markdown ? Juste donner un lien vers le fichiers et préciser dans un paragraphe que l'interface permet de manipuler un système de fichier ? A toi de voir mais en l'état, la liste n'apporte pas grand chose :)

> Vous pouvez remarquer que certaine méthodes sont suffixées par `Stream`, cela signifie manipuler un fichier en tant que [ressource](https://www.php.net/manual/en/language.types.resource.php). Dans le cas contraire vous manipuler un fichier sous forme de chaîne de caractère. 

Reviewer: Ah! `*Stream` est expliqué ici, my bad...

``

Un autre exemple d’utilisation. Le déplacement d’un fichier d’un système de fichier à un autre système de fichier. Pour ce faire nous aurons besoin d’utiliser le `MountManager` de Flysystem.

Le `MountManager` permet de faire des manipulation dans plusieurs système de fichier préalablement renseigner. Il suffira uniquement de préfixé le paramètre `path` avec le nom du système de fichier à utiliser.

Voici l’exemple pour déplacer un fichier d’un système de fichier local vers un Storage Google :

```php
// …
public function copToGcp(FilesystemInterface $localStorage, FilesystemInterface $gcpStorage)
    {
         $mountManager = new MountManager();
         $mountManager->mountFilesystems([
             'local' => $localStorage,
             'gcp' => $gcpStorage,
         ]);

         $mountManager->move('local://my_file.txt', 'gcp://mysdirectory/my_file_to_gcp.txt');
    }

// ...
```

Vous pouvez voir que l’on assigne un clé au système de fichier renseigner au `MountManager` comme ceci `'local' => $localStorage`. Ensuite, pour spécifier le système de fichier on prefix le `path` par la clé de précédemment défini `local://my_file.txt`.

> ## Un plugin pour customiser vos actions

Reviewer: préférer **personnaliser** que **customiser** 🇫🇷

{
    /** @var FilesystemInterface */
    protected $filesystem;

    public function setFilesystem(FilesystemInterface $filesystem): void
    {
        $this->filesystem = $filesystem;
    }

    public function getMethod(): string
    {
        return 'saveInvoice';
    }

    public function handle(int $id, string $content)
    {
        $now = new \DateTimeImmutable();
        $this->filesystem->write(
            sprintf('%s/%d-%s', $now->format('Y-m-d'), $id, $now->format('H-i-s')),
            $content
        );
    }
}
```

3. L’utiliser du plugin (Attention ici il y a du fake il faut réadapter pour une vraie utilisation)

```php
    public function saveInvoice(FilesystemInterface $defaultStorage, FilesystemInterface $invoiceStorage)
    {
        $invoice = $defaultStorage->read('invoice.txt');

        $invoiceStorage->addPlugin(new invoicePlugin());
        $invoiceStorage->saveInvoice($id, $invoice);
    }
```

4. Voir le résultat :

![result plugin Flysystem]({{ site.baseurl }}/assets/2019-11-20-reprenez-le-controle-de-vos-file-system-avec-flysystem/screenshot-result-plugin.png){: style="margin: 0 auto; display: block;" }

Et voilà pour ce dernier exemple.

> # En conclusion

Reviewer: `## En conclusion` titre de niveau 2.

Après je suis pas fan des titre du style "Introduction" et "Conclusion" car ça fait "trop scolaire", mais c'est une affaire de goût :)

---
layout: post
title: Reprenez le contrôle de vos filesystem avec Flysystem
excerpt: Cet article vous présente comment interagir simplement avec vos systèmes de gestion de fichier avec Flysystem.
authors:
    - nicolas
lang: fr
permalink: /fr/reprenez-le-controle-de-vos-file-system-avec-Flysystem/
categories:
    - PHP
    - Bundle
    - Filesystem
tags:
    - PHP
    - Symfony
    - Flysystem
    - Filesystem
---

La gestion d’un ou plusieurs système de fichier dans une application PHP peut s’avérer compliqué. Il est vrai qu'avec les fonctions natives il est assez simple de manipuler des fichiers en local. Si l'on regarde [la documentation officiel](https://www.php.net/manual/en/ref.filesystem.php) on peut faire un sytème rapidement comme ceci: 

```php
public function write(string $content, string $path): void
{
    if (file_exists($path)) {
      // throw exception
    }

    file_put_contents($path, $content);
}
```

> C'est faisable mais le jour où vous allez devoir sur un système de fichier distant comme AWS vous allez devoir tout recommencer. Il vous faudra installer le SDK AWS, le configurer puis re-développer votre fonction comme ceci:

Reviewer: ```suggestion
C'est faisable, mais le jour où vous allez migré sur un système de fichier distant comme AWS, vous allez devoir tout recommencer. Il vous faudra installer le SDK AWS, le configurer puis re-développer votre fonction comme ceci:
```

```php
use AWS\S3\S3Client

private s3Client;

public function __construct()
{
    $options = [
        'region' => 'us-west-2',
        'version' => '2006-03-01',
        'signature_version' => 'v4',
    ];

    $this->s3Client = new Aws\S3\S3Client([
        'region' => '-- your region --',
        'version' => 'latest',
        'credentials' => [
            'key' => "-- access key id --",
            'secret' => "-- secret access key --",
        ],
    ]);
}

public function write(string $content, string $path)
{
    $this->s3Client->putObject([
        'Bucket' => '-- bucket name --',
        'Key' => $path,
        'Body' => $content,
    ]);
}
```

> Et le jour ou l'on va vous demande de passer sur Google ? Ou bien qu'il faudra gérer un multitude de système de fichier ? Cela va être lourd à écrire et maintenir, sans parler des tests d'intégrations.

Reviewer: ```suggestion
Et si demain on vous demande de passer sur Google Cloud Storage ou bien de gérer un multitude de systèmes de fichiers ? Cela va être lourd à écrire et à maintenir, sans parler des tests d'intégrations...
```

> Pour eviter trop de complexité on pourrais utiliser le patron de conception [Adaptateur](https://fr.wikipedia.org/wiki/Adaptateur_(patron_de_conception)) et là on pourrais faire du code plus propre et plus maintenable. 

Reviewer: ```suggestion
Pour supporter plusieurs systèmes de fichiers on peut éventuellement mettre en œuvre le patron de conception [Adaptateur](https://fr.wikipedia.org/wiki/Adaptateur_(patron_de_conception)) ce qui nous permettrai d'abstraire les opérations sur les fichiers et de supporter plusieurs implémentations. 
```

> Le patron de conception `Adaptateur` permet de convertir l’interface d’un class une autre interface. Cela permet de faire fonctionner des classes qui n’aurait pas pu fonctionner ensemble.

Reviewer: ```suggestion
Le patron de conception `Adaptateur` permet de convertir l’interface d’une classe en une autre interface attendue par notre système. Cela permet de faire d'adapter des classes dont nous ne maitrisons pas la signature.
```

> Si l’on prend notre cas de notre système de gestion de système de fichier qui doit utiliser plusieurs clients ou librairies différentes (AWS, SFTP, fonction native) alors nous pourrons faire comme ceci :

Reviewer: ```suggestion
Si l’on prend le cas de notre système de gestion de système de fichiers qui doit utiliser plusieurs clients ou librairies différentes (AWS, SFTP, fonction native, etc...) alors nous pouvons faire comme ceci :
```

> - On créé une classe que l’on va instancier pour manipuler nos fichier :

Reviewer: ```suggestion
- On créé une classe que l’on va instancier pour manipuler nos fichiers :
```

```php
class FileSystem
{
    /** @var AdapterInterface */
    private $adapter;

    public function __construct(AdapterInterface $adapter)
    {
        $this->adapter = $adapter;
    }

    public function write(string $path, string $content, array $config): bool
    {
        return $this->adapter->write($path, $content, $config);
    }
}
```

> - Puis, on créé un interface qui va définir les méthodes qui devront être implémentés dans nos adaptateur : 

Reviewer: ```suggestion
- Puis, on créé une interface qui va définir les méthodes qui devront être implémentées dans notre adaptateur : 
```

```php
interface AdapterInterface
{
    public function write(string $path, string $content, array $config): bool;
}
```

> - Enfin, on créé nos x adaptateur :

Reviewer: ```suggestion
- Enfin, on créé nos _**n**_ adaptateurs :
```

    {
        print_r("write with local adapter\n");

        return true;
    }
}

class AWSAdapter implements AdapterInterface
{
    private s3Client;

    public function __construct()
    {
        $options = [
            'region' => 'us-west-2',
            'version' => '2006-03-01',
            'signature_version' => 'v4',
        ];

        $this->s3Client = new Aws\S3\S3Client([
            'region' => '-- your region --',
            'version' => 'latest',
            'credentials' => [
                'key' => "-- access key id --",
                'secret' => "-- secret access key --",
            ],
        ]);
    }

    public function write(string $path, string $content, array $config): bool
    {
        print_r("write with aws adapter\n");

        return true;
    }
}
```

Et voilà !

Si on test notre code ca donne ceci :

```php
$localAdapter = new LocalAdapter();
$localFileSystem = new FileSystem($localAdapter);
$localFileSystem->write('', '', []);

$awsAdapter = new AWSAdapter();
$awsFileSystem = new FileSystem($awsAdapter);
$awsFileSystem->write('', '', []);
```

```bash
$ php index.php
write with local adapter
write with aws adapter
```

> On voit comme nous avons fait abstraction des opérations du système de fichier et on vient adapter les interfaces de chacune des librairies que l'on souhaite utiliser.

Reviewer: ```suggestion
Nous avons créé une abstraction pour manipuler un système de fichiers puis nous avons adapter les interfaces de chacune des librairies que l'on souhaitait utiliser.
```

> Mais bon on doit encore faire du code. En plus en librairie exploite déjà les système de gestion de fichier comme ça.

Reviewer: ```suggestion
Mais bon, on doit encore maintenir du code. Heureusement pour nous, il existe déjà une librairie qui s'en charge déjà.
```

> Je vous persente [Flysystem](https://flysystem.thephpleague.com/docs/) développer par [thephpleague](https://thephpleague.com/fr/), un groupe de développer de bibliothèques PHP. *Flysystem* est une bibliothèque d'abstraction du système de fichier. 

Reviewer: ```suggestion
Je vous persente [Flysystem](https://flysystem.thephpleague.com/docs/) développer par [thephpleague](https://thephpleague.com/fr/), un groupe de développeur de bibliothèques PHP. *Flysystem* est une bibliothèque d'abstraction du système de fichiers. 
```

> Cela permet donc de changer du solution de système de fichier rapidement et facilement grâce au patron de conception `Adaptateur`. 

Reviewer: ```suggestion
Cela permet donc de changer du solution de système de fichiers rapidement et facilement grâce au patron de conception `Adaptateur`. 
```

Vous pouvez l’utiliser dans une application PHP avec ou sans framework.

> Flysystem fourni une API permettant de gérer vos ressources sur un grand nombre de système de fichier. D’office, la librairie fournit trois adaptateurs de système de fichier, FTP, Local et NullAdapter. Mais rien ne vous empêche d’ajouter d’autre adaptateur de système de fichier, d'ailleurs il en existe un grand nombre.

Reviewer: ```suggestion
Flysystem fourni une API permettant de gérer vos ressources sur un grand nombre de système de fichiers. D’office, la librairie fournit trois adaptateurs de système de fichier, FTP, Local et NullAdapter. Mais rien ne vous empêche d’ajouter d’autre adaptateur de système de fichier, d'ailleurs, il en existe un grand nombre.
```

Vous pouvez retrouver la liste complète sur [le README du dépôt officiel](https://github.com/thephpleague/flysystem). Si par malheur vous ne trouvez pas votre bonheur parmi la liste proposée vous pouvez développer le vôtre. Voici le lien pour [créer un adaptateur](https://flysystem.thephpleague.com/docs/advanced/creating-an-adapter/) car je n’en parlerait pas ici.  

## Permuter de système de fichier avec Flysystem

Je vais vous montrer un exemple de permutation de système de fichier avec Flysystem dans une application Symfony 4.

> Dans un premier temps il faut déjà configurer un système de fichier avec Flysystem. Pour ce faire nous allons installer le bundle avec composer comme ceci `composer require league/flysystem-bundle`, puis on édite la configuration de Flysystem dans le fichier `config/packages/flysystem.yaml` avec ces quelques lignes :

Reviewer: ```suggestion
Dans un premier temps, il faut déjà configurer un système de fichiers avec Flysystem. Pour ce faire, nous allons installer le bundle avec composer comme ceci : `composer require league/flysystem-bundle`, puis on édite la configuration de Flysystem dans le fichier `config/packages/flysystem.yaml` avec ces quelques lignes :
```

```yaml
flysystem:
  storages: 
    default.storage:
      adapter: 'local'
      options:
        directory: '%kernel.project_dir%/var/storage'
```

> Ici on peut voir que l’on configure un système de fichier qui ce nome `default.storage`, utilisant l’adaptateur `local` et qu’il y a une option qui cible le répertoire ou seront stockés les ressources: `directory: '%kernel.project_dir%/var/storage'`.

Reviewer: ```suggestion
Ici on peut voir que l’on configure un système de fichier qui ce nomme `default.storage`, utilisant l’adaptateur `local` et qu’il y a une option qui cible le répertoire où seront stockés les ressources: `directory: '%kernel.project_dir%/var/storage'`.
```

> Jusqu'ici rien de compliqué. Mais maintenant on change de système de fichier pour passer sur une solution de stockage sur AWS S3. Pour ce faire nous allons dans un premier temps [créer un Bucket dans la console AWS](https://docs.aws.amazon.com/fr_fr/AmazonS3/latest/user-guide/create-bucket.html). Puis, nous installons l'adaptateur Flysystem AWS S3 via composer comme ceci `composer require league/flysystem-aws-s3-v3`.

Reviewer: ```suggestion
Jusqu'ici rien de bien compliqué. Mais maintenant, on change de système de fichiers pour passer sur une solution de stockage sur AWS S3. Pour ce faire, nous allons dans un premier temps [créer un Bucket avec la console AWS](https://docs.aws.amazon.com/fr_fr/AmazonS3/latest/user-guide/create-bucket.html). Puis, nous installons l'adaptateur Flysystem AWS S3 via composer comme ceci `composer require league/flysystem-aws-s3-v3`.
```

Ensuite nous configurons notre client AWS :

```yaml
services:
# ...
  Aws\S3\S3Client:
    arguments:
        - version: 'latest'
          region: '%env(string:AWS_REGION)%'
          credentials:
            key: '%env(string:AWS_CREDENTIALS_KEY)%'
            secret: '%env(string:AWS_CREDENTIALS_SECRET)%'
          S3:
            version: "2006-03-01"
```

> Il faudra récupérer des `credentials` sur la console AWS pour ça je vous laisse regarder la [documentation officiel sur AWS](https://docs.aws.amazon.com/fr_fr/cli/latest/userguide/cli-chap-configure.html).

Reviewer: ```suggestion
Il faudra récupérer des `credentials` sur la console AWS, pour ce faire, je vous laisse consulter la [documentation officiel sur AWS](https://docs.aws.amazon.com/fr_fr/cli/latest/userguide/cli-chap-configure.html).
```

> Une fois le client configurer, il nous reste à permuter de système de fichier. On retourne dans `config/packages/flysystem.yaml` et modifie quelques lignes comme ceci :

Reviewer: ```suggestion
Une fois le client configuré, il nous reste à permuter de système de fichiers. On retourne dans `config/packages/flysystem.yaml` et on modifie quelques lignes comme ceci :
```

```yaml
flysystem:
  storages:
    default.storage:
      adapter: 'aws'
      options:
        client: Aws\S3\S3Client
        bucket: '%env(string:AWS_BUCKET)%'
```

> Et voilà on viens de changer de système de fichier juste en changeant la valeur de l’adaptateur de `local` à `aws` et en modifiant les options pour spécifier le client et le bucket à utiliser.

Reviewer: ```suggestion
Et voilà, on vient de changer de système de fichiers juste en changeant la valeur de l’adaptateur de `local` à `aws` et en modifiant les options pour spécifier le client et le bucket à utiliser.
```

> ##  Utiliser plusieurs system de fichier avec Flysystem

Reviewer: ```suggestion
##  Utiliser plusieurs system de fichiers avec Flysystem
```

> On a vu comment permuter facilement de système de fichier, maintenant nous allons voir comment configurer plusieurs système de fichier. Je vais vous faire un exemple de configuration avec trois solution de stockage. Un système de fichier locale et deux solution cloud (AWS et GCP). 

Reviewer: ```suggestion
On a vu comment permuter facilement de système de fichiers, maintenant nous allons voir comment configurer plusieurs système de fichiers. Je vais vous faire un exemple de configuration avec trois solution de stockage. Un système de fichiers local et deux solutions cloud (AWS et GCP). 
```

> La première étage va être d’installer le bundle et l'adaptateur pour AWS S3 , ainsi que l'adaptateur pour Google Storage. `composer require league/flysystem-bundle league/flysystem-aws-s3-v3 superbalist/flysystem-google-storage`.

Reviewer: ```suggestion
La première étape va consiter à installer le bundle et l'adaptateur pour AWS S3 , ainsi que l'adaptateur pour Google Storage. `composer require league/flysystem-bundle league/flysystem-aws-s3-v3 superbalist/flysystem-google-storage`.
```

> Maintenant nous allons configurer les clients AWS et GCP. Pour AWS pas de changement, il faire ce que l’on a fait pour la permutation. 

Reviewer: ```suggestion
Maintenant, nous allons configurer les clients AWS et GCP. Pour AWS pas de changement.
```

`il faire ce que l’on a fait pour la permutation.` je pense qu'il manque un mot car ça n'a pas de sens :)

> Pour GCP, il faut [créer un storage](https://cloud.google.com/storage/docs/creating-buckets?hl=fr) et récupère un [fichier json d’authentification ](https://cloud.google.com/video-intelligence/docs/common/auth?hl=fr#set_up_a_service_account).

Reviewer: ```suggestion
Pour GCP, il faut [créer un storage](https://cloud.google.com/storage/docs/creating-buckets?hl=fr) et récupérer un [fichier json d’authentification](https://cloud.google.com/video-intelligence/docs/common/auth?hl=fr#set_up_a_service_account).
```

> Une fois que l’on a configurer le Bucket AWS S3 ainsi que le Storage Google et que l’on a récupérer les information de connection il nous reste plus qu'à configurer les clients comme ceci :

Reviewer: ```suggestion
Une fois que l’on a configurer le Bucket AWS S3 ainsi que le Storage Google et que l’on a récupéré les information de connexion, il ne nous reste plus qu'à configurer les clients comme ceci :
```

```yaml
# config/services.yaml
services :
# …

  Google\Cloud\Storage\StorageClient:
    arguments:
      - keyFilePath: '%env(string:GCP_AUTH_FILE)%' #fichier json contenant les information de connection

  Aws\S3\S3Client:
    arguments:
        - version: 'latest'
          region: '%env(string:AWS_REGION)%'
          credentials:
            key: '%env(string:AWS_CREDENTIALS_KEY)%'
            secret: '%env(AWS_CREDENTIALS_SECRET)%'
          S3:
            version: "2006-03-01"
```

Maintenant on va configurer nos système de fichier :

```yaml
flysystem:
    local.storage:
      adapter: 'local'
      options:
        directory: '%kernel.project_dir%/var/storage'

    gcp.storage:
      adapter: 'gcloud'
      options:
        client: Google\Cloud\Storage\StorageClient
        bucket: '%env(string:GCP_BUCKET)%'
        api_url: 'https://storage.googleapis.com'

    aws.storage:
      adapter: 'aws'
      options:
        client: Aws\S3\S3Client
        bucket: '%env(string:AWS_BUCKET)%'
```

> Alors dans c’est exemple je nomme mes systèmes de fichier par rapport à leurs adaptateur mais vous pouvez les nommer par rapport à leur utilité exemple :

Reviewer: ```suggestion
Alors, dans ces exemples, je nomme mes systèmes de fichiers en fonction du nom de l'adaptateur, mais vous pouvez les nommer en fonction de leur utilité, exemple :
```

```yaml
flysystem:
    assets.storage:
      adapter: 'local'
      # ...

    invoice.storage:
      adapter: 'gcloud'
      # ...

    media.storage:
      adapter: 'aws'
      # ...
``` 

> Et voilà pour la configuration de plusieurs système de fichier.

Reviewer: ```suggestion
Et voilà pour la configuration de plusieurs système de fichiers.
```

## La configuration ça va deux minutes

> Nous avons vu comment permuter de système de fichier juste avec quelque lignes de configuration, mais aussi comment configurer plusieurs système de fichier.

Reviewer: ```suggestion
Nous avons vu comment permuter de système de fichiers juste avec quelque lignes de configuration, mais aussi comment configurer plusieurs système de fichiers.
```
